esteam script to install dependencies, plus general clean up

chewi opened the pull request:
Introducing esteam! I have already spoken to @anyc about it but for the benefit of everyone else, this is a script designed to complement or perhaps eventually replace the existing steam-games-meta package. It automatically locates Steam libraries on your system and scans them with Gentoo's own scanelf in order to write a list of dependencies into a Portage set.

It does this with the help of mappings from library names to packages. These mappings are not yet complete but they are sufficient for the 12 games I currently have installed. Ideally we would cover as many of the libraries in the Steam runtime as we can and I am working my way through them, though some are much more likely to be needed than others. I don't imagine many games require libldap-2.4.so.2, for example.

This approach gives us tons more coverage of both games and libraries than the steam-games-meta package ever could because we only have to write one set of mappings for all the libraries and not a different set for each and every game. It also requires far less effort to use. Just install your games as usual, run esteam as root, let Portage do its thing, and you're done.

esteam will even delete bundled libraries in favour of system versions for a specific list of games where this has been deemed safe. This list will always exclude VAC-enabled games as deleting libraries probably still counts as modifying them. I have only seen this break with Hammerwatch as our system version of Mono does not appear to be compatible.

If deleting files sounds scary, fear not as Steam's _Verify Integrity of Game Cache_ will restore them very quickly. The feature is not currently optional but if enough people dislike it then that could be changed. For what it's worth, relatively few free software libraries are bundled as most of them are included in the Steam runtime.

The deletion feature can also get us out of tricky spots like Alien Isolation. That game bundles a version of libcurl.so.4 that links against librtmp.so.0. Our tree only has librtmp.so.1 but by using our system version of libcurl.so.4, the problem disappears. I haven't yet determined whether this game needs the RTMP support but it should be possible to make the USE flags for some of the packages a little more dynamic.

I've tried to ensure that esteam covers all the aspects dealt with by steam-games-meta, even including the detection of xwininfo in Trine 2's launcher script. Any files beginning with `#!` are scanned for the presence of xwininfo and potentially other utilities in future so the detection is not limited to just libraries. This shebang scan is relatively fast, thanks to some find/awk/xargs trickery.

Adobe AIR and S3TC are dealt with as special cases. The former should be accurate for all AIR-based games but has only been tested with Anodyne. S3TC is hard to detect accurately. If you grep for `GL_EXT_texture_compression_s3tc` then almost every game matches, including ones that blatantly don't use it like 1001 Spikes. The literal uppercase string `S3TC` on the other hand gives much better results. You only need it to match one game for the package to be included and it only gets included for Mesa users (based on eselect opengl) anyway. Worst case, you can just install that package manually.

Are you a PulseAudio hater? Don't worry, I've got you covered. Disabling the pulseaudio flag on the esteam package will make it swap media-sound/pulseaudio for media-sound/apulse in the package mappings. It still won't add anything to `LD_LIBRARY_PATH` for you though.

You can use esteam with the Steam runtime enabled (via the USE flag or `STEAM_RUNTIME=1`) but it will still needlessly install libraries already covered by the runtime and they won't even be used. This is probably only useful for 64-bit games as Steam doesn't ship a 64-bit runtime. I'm not very interested in enabling the runtime but if enough people speak up, I could make esteam optionally install just the packages that the runtime doesn't cover. I could possibly even make esteam delete libraries from the runtime, leaving just the ones that we don't have packaged. This would give the best of both worlds.

Probably the biggest caveat right now is that esteam will install 32-bit versions of packages for games that are also 64-bit. This doesn't matter much to me as I enable `abi_x86_32` globally but if you don't, this could be annoying. With sufficient interest, I could make esteam ignore the 32-bit binaries if 64-bit binaries are also present. This could potentially break if, for example, you have a 32-bit game with a 64-bit level editor. That seems quite unlikely though.

Dealing with this issue would also nobble some harmless error messages that can appear. I've only seen this one with game, A Boy and His Blob, so I don't know how common it is but sometimes 64-bit games include the 32-bit executables but not the 32-bit libraries required to support them.

esteam includes some help text that you can view with `-h`, although most of that information is included above. There is also a verbose `-v` option that is useful in finding possible additions to the unbundleable list. If you find that a VAC-free game works after unbundling (i.e. you've done more than just fire it up) then please file a pull request!

While working on this, I cleaned up the entire overlay. This included adding 1.6 and 1.10 SLOTs to the main tree's media-libs/glew so that media-libs/steam-runtime-glew could be dropped. I also added dev-libs/libpcre-debian as a general solution to [the libpcre.so.3 problem](https://archives.gentoo.org/gentoo-dev/message/676cd8b16a7255446355744d91636880). Being a Gentoo dev, I can easily push clean solutions like this and would rather do so than add Steam-specific fixes here.

I'm not looking to get this merged immediately. I'll wait for some feedback to ensure that you guys are broadly happy with it. If you like what you see then please let us know!

Some notes about specific games:

### Hammerwatch

steam-games-meta pulls in Mono but only for the editor as the game itself now bundles Mono. I don't understand why the bundled Mono cannot be used to launch the editor and I couldn't launch it either way. This may be because our Mono is incompatible, as I mentioned earlier.

### The Witcher 2

I don't know whether #107 is still an issue but I'd like to find out. I own this game but only on GOG, not Steam. Chances are the build is more or less the same so it'd be worth a try but I also don't have it downloaded yet. My broadband is capped so I'll download it at the end of the month if I have enough quota remaining, which I probably will.

### Crusader Kings II

dev-libs/steam-runtime-openssl was added in #151 to try and fix this but the fix was never verified. esteam doesn't pull this in and I'd like to know if there's a better way to fix it but I'm not interested in buying it. Please speak up if you have it.

chewi commented:
I have just made a big push to get many more entries into the mapping database that correspond to libraries in the Steam runtime. The remaining libraries would require significant packaging effort.  We may not need them all though.

One package in particular bothers me but hopefully it won't be an issue in practise. Some libraries are provided by mit-krb5 and some are provided by heimdal but these packages conflict with each other. Ubuntu doesn't have this problem because it splits up the packages a lot more. I doubt any game would depend on any of these libraries but if this does become a problem, I'd either have to split heimdal up or add a USE flag to control the installation of krb5 libraries.

anyc commented:
Thanks, it looks good so far to me!

In case you're interested, I only had to add CONSOLETYPE="foo" before sourcing the gentoo functions (and comment the truncate and emerge command) to run your script as a regular user. Without CONSOLETYPE, the functions script will abort the whole script.

The warnings I get for the installed games:
```
 * Unknown: libcurl-gnutls.so.4 needed by dont_starve/bin/dontstarve_steam
 * Unknown: libtier0_srv.so needed by Counter-Strike Source/cstrike/bin/server_srv.so
 * Unknown: libvstdlib_srv.so needed by Counter-Strike Source/cstrike/bin/server_srv.so
 * Unknown: libopenvr_api.so needed by OpenVR/bin/linux32/vrcmd_linux
 * Unknown: libMiles.so needed by Nuclear Dawn/bin/vaudio_miles.so
 * Unknown: libtiff.so.4 needed by Dwarfs - F2P/libgdiplus.so.0
 * Unknown: libxcb.so.1 needed by Dwarfs - F2P/libX11.so.6
 * Unknown: libgnutls.so.26 needed by Dwarfs - F2P/libcups.so.2
 * Unknown: libmad.so.0 needed by Dwarfs - F2P/libSDL_mixer-1.2.so.0
 * Unknown: libSteamValidateUserIDTickets_amd64.so needed by Half-Life/engine_amd64.so
 * Unknown: libstdc++.so.5 needed by Half-Life/engine_amd64.so
 * Unknown: libjawt.so needed by Altitude/native/liblwjgl64.so
 * Unknown: libjvm.so needed by Altitude/native/libduath.speex_x86_64.so
 * Unknown: libthread_db.so.1 needed by Altitude/jre/lib/i386/libsaproc.so
 * Unknown: libavcodec.so.52 needed by Altitude/jre/lib/i386/fxavcodecplugin-52.so
 * Unknown: libavformat.so.52 needed by Altitude/jre/lib/i386/fxavcodecplugin-52.so
 * Unknown: libavcodec.so.53 needed by Altitude/jre/lib/i386/fxavcodecplugin-53.so
 * Unknown: libavformat.so.53 needed by Altitude/jre/lib/i386/fxavcodecplugin-53.so
 * Unknown: libnsl.so.1 needed by Altitude/jre/bin/javaws
```

chewi commented:
Thanks for the feedback! Out of interest, how many games do you have installed?

I am aware of the CONSOLETYPE issue but didn't prioritise it as I didn't intend for the script to be run as non-root. It may try to delete libraries from games installed by other users and of course, it needs to run emerge at the end. I don't blame you for not wanting to run it as root though, at least initially. The issue is a bug in gentoo-functions itself as consoletype is in sbin (not in the user PATH) and it can fail when run as non-root anyway. Neither of these should stall a script including the functions so I'll file a bug soon.

I have Half-Life so I've installed that now. I have Counter-Strike but not the Source version unfortunately. Altitude and Dwarfs are free so I've now installed those.

First, the easy bits. libnsl.so.1 and libthread_db.so.1 are part of glibc so I've added those. It's strange that libmad.so.0 isn't bundled as it doesn't appear to be in the runtime. We have it packaged though so I've added that. I didn't add libtiff.so.4 before because I didn't realise our tiff package was slotted. Now we can handle versions 3, 4, and 5!

I do own Nuclear Dawn but don't particularly want to download it. libMiles.so should probably be bundled as it's not one we can provide. vrcmd_linux seems to be related to VR and therefore not critical to the game. I'm starting to get the impression that we'll see a lot of errors like this that we can do nothing about.

libstdc++.so.5 is available in the legacy libstdc++-v3 package so I've added it for completeness but I'm not sure why engine_amd64.so is present in Half-Life at all. It's the only 64-bit binary in a 32-bit game and a prime example of why we cannot blindly ignore 32-bit binaries when 64-bit ones are present. I've read that there was a 64-bit dedicated server but installing the dedicated server through Steam didn't give me any 64-bit executables. Just another anomaly to add to the list, I suppose.

Anything XCB-related is going to be a pain as it's never ABI-compatible from one version to the next. GnuTLS is another tricky one as the ABI changes a lot. The good news is that adding Dwarfs to the unbundleables list nips these two in the bud as libX11 and libcups get deleted.

Dwarfs still wouldn't start though, even with the Steam runtime enabled. Does it start for you? Searching revealed that [deleting libtier0_s.so and libvstdlib_s.so does the trick](https://steamcommunity.com/app/213650/discussions/0/864969320369614550/?ctp=2#c864977025817087468). I was in two minds about whether to have esteam handle this specific oddity. Nothing else is going to fix it so I added this as a generic DELETEABLES feature that will simply delete the given files if they're found. Now it works out of the box.

I'm not sure what to do about Don't Starve. curl can currently only be built against one backend. I recall hacking another game, Funky Smugglers, to just point at the regular libcurl.so.4 and it worked fine. A packaged symlink may therefore be the answer. If that doesn't work, we'd have to repackage curl to install a GnuTLS-backed version in /usr/lib/debiancompat.

Altitude has packaged the Oracle JRE, which is not only legally questionable but also bloody stupid, not least because of the non-stop conveyor belt of Java vulnerabilities. This obviously makes esteam complain loudly. It's requesting a version of ffmpeg that we don't have and it's not even the version that the runtime has. That doesn't matter though, these errors stem from JavaFX libraries that the game doesn't even use. Being the Gentoo Java lead, I feel strongly that we should just unbundle the whole JRE to allow safe and free alternatives to be used. This game does include 64-bit libraries but it's bundled a 32-bit JRE and I think it would generally be safer to replace like with like. Luckily I added multilib to icedtea-bin very recently. I'll think about how I can make this work.

As well as the fixes already mentioned, I have dropped Mono from the mappings because it's not multilib-enabled and unlikely to work as a replacement anyway. I also fixed some x11-libs packages that had the wrong category.

Tele42 commented:
Hello @chewi, specific to Don't Starve Together, it does not appear to care which backend is used with curl. I personally resolved this dependency by replacing the bundled library with a symlink to the system version in the game's lib32 folder.

I really wish steam had post-update and post-install hooks for package maintainers.

anyc commented:
Oh, I forgot your question after I read your post :) About 30 games are installed. One played in the last 2 years or so. ;)

So, as nobody complains, shall we merge it?

chewi commented:
As you can see, I just pushed an update to allow it to be run as non-root. It obviously needs to call sudo at the end but the scan itself works as a regular user as long as that user has write permission to the Steam libraries. If other users have Steam installed then it probably won't even see those games so the scan will be incomplete.

I did put the word out on the Gentoo forums and informed a couple of users on the Phoronix forums but I don't believe anyone else has tried it yet. Following @Tele42's feedback, I'd like to deal with GnuTLS using a packaged symlink. I'll keep that package in this overlay for now as I'm a little unsure about it. Let's merge it after that.

It would be handy to do updates without filing pull requests so commit access here would be appreciated. Up to you though, of course.

chewi commented:
Many thanks, @anyc. :smile:

anyc commented:
Np, you're doing more work than me lately. ;)

chewi commented:
I'll update the README soon to include this new method.
